src/coreclr/jit/codegenarm64.cpp

```diff
@@ -3996,6 +3996,12 @@ void CodeGen::genCodeForCmpXchg(GenTreeCmpXchg* treeNode)
             ins = INS_casalh;
         }
         GetEmitter()->emitIns_R_R_R(ins, dataSize, targetReg, dataReg, addrReg);
+
+        if (varTypeIsSmall(treeNode->TypeGet()) && varTypeIsSigned(treeNode->TypeGet()))
+        {
+            instruction mov = varTypeIsShort(treeNode->TypeGet()) ? INS_sxth : INS_sxtb;
+            GetEmitter()->emitIns_Mov(mov, EA_4BYTE, targetReg, targetReg, /* canSkip */ false);
```

Comment:
What is this mov for?

Reply:
Ah, I misread the code, this is just the extension.

Reply:
Gonna answer anyway: `casalb/h` and `ldaxrb/h` always zero extend the value, we need to force sign extend it for signed types, the issue here was that we sign extended it after compare in the 8.0 path, not before it which broke with negative comparands.
This could be fixed by force zero extending the comparand but that seems trickier to do in the JIT, we'd probably need to force cast it to unsigned in the importer.